File: Android/TagMaster/src/main/java/depollsoft/tagmaster/ListModel.kt
package depollsoft.tagmaster

import com.google.firebase.Firebase
import com.google.firebase.auth.auth
import com.google.firebase.firestore.FieldPath
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.ListenerRegistration
import com.google.firebase.firestore.MetadataChanges
import com.google.firebase.firestore.SetOptions
import com.google.firebase.firestore.firestore
import depollsoft.lib.state.StateList
import depollsoft.lib.util.Preferences

/**
 * One list of tag ids, kept in [Preferences] and, when someone is signed in, in the `lists` map of
 * their Firestore user document.
 *
 * [ids] is snapshot state, so composables showing a list recompose when it changes. Every change
 * goes through this class, which stores it on the spot: the list is saved before the call that
 * changed it returns.
 */
class ListModel private constructor(
    val listName: String,
) {
    private val store: StateList<Int> = preferences[listName] ?: StateList()

    /** The tags in this list, in order. Assigning replaces the contents as one change. */
    var ids: List<Int>
        get() = store
        set(value) {
            val next = value.toList()
            store.replaceWith(next)
            changed()
        }

    private var revision = 0L

    /** A main-thread editing baseline; the revision detects any change made since it was taken. */
    class Snapshot internal constructor(
        internal val revision: Long,
        val ids: List<Int>,
    )

    fun snapshot() = Snapshot(revision, store.snapshot())

    /** Commit a complete permutation once, only if nothing changed since editing began. */
    fun reorder(
        expected: Snapshot,
        order: List<Int>,
    ): Boolean {
        if (revision != expected.revision || store.snapshot() != expected.ids) return false
        if (order == expected.ids || order.size != expected.ids.size ||
            order.toSet().size != order.size || order.toSet() != expected.ids.toSet()
        ) {
            return false
        }
        store.replaceWith(order)
        changed()
        return true
    }

    /** Move by stable tag ID, rejecting missing IDs, out-of-range destinations and no-ops. */
    fun move(
        id: Int,
        destination: Int,
    ): Boolean {
        val baseline = snapshot()
        val from = baseline.ids.indexOf(id)
        if (from < 0 || destination !in baseline.ids.indices || from == destination) return false
        val order = baseline.ids.toMutableList()
        order.add(destination, order.removeAt(from))
        return reorder(baseline, order)
    }

    fun add(id: Int) {
        if (store.contains(id)) return
        store.add(id)
        changed()
    }

    fun canMoveDown(id: Int): Boolean {
        val index: Int = store.indexOf(id)
        return index >= 0 && index < store.size - 1
    }

    fun canMoveUp(id: Int): Boolean {
        val index: Int = store.indexOf(id)
        return index > 0
    }

    fun contains(id: Int): Boolean = store.contains(id)

    fun moveDown(id: Int) {
        if (canMoveDown(id)) move(id, store.indexOf(id) + 1)
    }

    fun moveUp(id: Int) {
        if (canMoveUp(id)) move(id, store.indexOf(id) - 1)
    }

    fun remove(id: Int) {
        if (store.remove(Integer.valueOf(id))) changed()
    }

    fun reset() {
        if (store.isEmpty()) return
        store.clear()
        changed()
    }

    /** Replaces the contents with a cloud copy; stored locally only, as [fromFirestore] requires. */
    private fun replaceFromRemote(value: List<Int>) {
        if (value == store.snapshot()) return
        store.replaceWith(value)
        changed()
    }

    private fun changed() {
        revision++
        storeValue()
    }

    private fun storeValue() {
        if (store.isEmpty()) {
            preferences.remove(listName)
        } else {
            preferences[listName] = store
        }
        if (!shouldStore) {
            return
        }
        Companion.storeValue(false)
        val user = Firebase.auth.currentUser
        if (user != null) {
            val userDoc = Firebase.firestore.document("users/${user.uid}")
            userDoc.set(
                mapOf(
                    "lists" to
                        mapOf(
                            listName to if (store.isEmpty()) FieldValue.delete() else store.toList(),
                        ),
                ),
                // A FieldPath keeps list keys with dots or other punctuation intact.
                SetOptions.mergeFieldPaths(listOf(FieldPath.of("lists", listName))),
            )
        }
    }

    companion object {
        private const val LISTS_KEY = "tagmaster.lists"
        private const val SYNCED_UID_KEY = "tagmaster.syncedUid"

        private val preferences: MutableMap<String, StateList<Int>> by lazy {
            decodeLists(Preferences.get(LISTS_KEY))
        }

        /**
         * Every stored list as the observable list the models use. Stored lists come back as
         * [StateList] (the app registers it under the alias the Bindroid collection was written
         * with); anything else that is a collection is converted rather than dropped.
         */
        @Suppress("UNCHECKED_CAST")
        internal fun decodeLists(stored: Map<*, *>?): MutableMap<String, StateList<Int>> {
            val lists = mutableMapOf<String, StateList<Int>>()
            stored?.forEach { (key, value) ->
                if (key is String && value is Collection<*>) {
                    lists[key] = value as? StateList<Int> ?: StateList(value.filterIsInstance<Int>())
                }
            }
            return lists
        }

        private fun storeValue(toFirestore: Boolean) {
            Preferences.setAsync(LISTS_KEY, preferences)
            if (toFirestore) {
                toFirestore()
            }
        }

        // Held strongly: a model is tiny, there is one per list, and composables read its `ids`.
        // A weak cache let an empty list (no stored preference) be rebuilt around a fresh
        // collection once the old model was collected, silently orphaning those readers.
        private val modelInstances: MutableMap<String, ListModel> = mutableMapOf()

        operator fun invoke(listName: String): ListModel = modelInstances.getOrPut(listName) { ListModel(listName) }

        private fun migrateOldFavorites() {
            val favoritesKey = "tagmaster.Favorites"
            val teachablesKey = "tagmaster.TeachableTags"

            if (Preferences.get<Any?>(LISTS_KEY) != null) {
                Preferences.set(favoritesKey, null)
                Preferences.set(teachablesKey, null)
                return
            }

            val favorites: Collection<*>? = Preferences.get(favoritesKey)
            val teachables: Collection<*>? = Preferences.get(teachablesKey)

            if (favorites != null) {
                ListModel.invoke(TagLists.FAVORITE).ids = favorites.filterIsInstance<Int>()
                Preferences.set(favoritesKey, null)
            }

            if (teachables != null) {
                ListModel.invoke(TagLists.TEACHABLE).ids = teachables.filterIsInstance<Int>()
                Preferences.set(teachablesKey, null)
            }
        }

        internal var shouldStore = true
            private set

        /** The keys of every list with at least one tag on this device. */
        internal fun storedKeys(): Set<String> = preferences.keys.toSet()

        /**
         * Drops every list on this device without touching the cloud. Used when a different
         * account signs in: the lists here belong to the account that last synced them.
         */
        internal fun forgetLocalLists() {
            val wasStoring = shouldStore
            shouldStore = false
            try {
                (preferences.keys + TagLists.RESERVED).toSet().forEach { ListModel.invoke(it).reset() }
                preferences.clear()
                TagLists.resetLocal()
            } finally {
                shouldStore = wasStoring
            }
            storeValue(false)
        }

        /**
         * Records which account this device's lists belong to. Signing in as a different account
         * than the one that last synced here drops the local lists first, so one user's lists are
         * never uploaded into another user's brand-new document. Signing out keeps the lists, and
         * the same account signing back in finds them untouched. Returns whether lists were dropped.
         */
        internal fun prepareLocalLists(forUid: String): Boolean {
            val synced = Preferences.get<String>(SYNCED_UID_KEY)
            Preferences.set(SYNCED_UID_KEY, forUid)
            if (synced == null || synced == forUid) return false
            forgetLocalLists()
            return true
        }

        /**
         * Empties [listName] on this device without writing to Firestore. [TagLists.delete] uses it
         * and then deletes the cloud field together with the list's metadata in one write.
         */
        internal fun discard(listName: String) {
            val wasStoring = shouldStore
            shouldStore = false
            try {
                ListModel.invoke(listName).reset()
            } finally {
                shouldStore = wasStoring
            }
            preferences.remove(listName)
            storeValue(false)
        }

        /**
         * For testing: disable Firebase storage to allow unit testing without Firebase initialization.
         * Call this in @Before methods of tests that create ListModel instances.
         */
        @JvmStatic
        fun setTestMode(enabled: Boolean) {
            shouldStore = !enabled
        }

        private fun fromFirestore(
            data: Map<*, *>,
            info: Map<*, *>?,
        ) {
            // Restored rather than forced on: a snapshot applied while storing is off (test mode)
            // must not switch cloud writes back on behind the caller's back.
            val wasStoring = shouldStore
            shouldStore = false
            try {
                // Remove lists that aren't in the data
                preferences.keys.subtract(data.keys).forEach {
                    if (it !is String) {
                        return
                    }
                    ListModel.invoke(it).reset()
                }
                // Update existing lists
                data.keys.forEach {
                    if (it !is String) {
                        return
                    }
                    val newValue =
                        (data[it] as? List<*>)
                            ?.mapNotNull { (it as? Long)?.toInt() }
                            ?: emptyList()
                    ListModel.invoke(it).replaceFromRemote(newValue)
                }
                TagLists.applyRemote(info, data.keys.filterIsInstance<String>())
            } finally {
                shouldStore = wasStoring
            }
            storeValue(false)
        }

        private fun toFirestore() {
            val user = Firebase.auth.currentUser
            if (user != null) {
                val userDoc = Firebase.firestore.document("users/${user.uid}")
                userDoc.set(
                    mapOf(
                        "lists" to preferences.mapValues { it.value.toList() },
                        "listInfo" to TagLists.remoteInfo(),
                    ),
                    SetOptions.merge(),
                )
            }
        }

        var registration: ListenerRegistration? = null

        fun connectToFirestore() {
            registration?.remove()
            val user = Firebase.auth.currentUser
            if (user != null) {
                prepareLocalLists(forUid = user.uid)
                val userDoc = Firebase.firestore.document("users/${user.uid}")
                // Metadata changes are included so the server's confirmation of a missing
                // document arrives even when nothing else changed; see applyUserSnapshot.
                registration =
                    userDoc.addSnapshotListener(MetadataChanges.INCLUDE) { snapshot, error ->
                        if (error != null) {
                            print(error)
                            return@addSnapshotListener
                        }
                        applyUserSnapshot(
                            exists = snapshot!!.exists(),
                            fromCache = snapshot.metadata.isFromCache,
                            lists = snapshot.get("lists") as? Map<*, *>,
                            info = snapshot.get("listInfo") as? Map<*, *>,
                        )
                    }
            }
        }

        /**
         * One user-document snapshot after sign-in.
         *
         * The account's document is the truth: its lists replace whatever this device had, so
         * signing in never carries local-only lists into an existing account. The one time this
         * device's lists are uploaded is when the server says the account has no document yet,
         * which is a brand-new account. A cache miss says nothing about the account (it is what an
         * offline start looks like), so it neither seeds the document nor clears the local lists;
         * the server-confirmed snapshot that follows decides.
         */
        internal fun applyUserSnapshot(
            exists: Boolean,
            fromCache: Boolean,
            lists: Map<*, *>?,
            info: Map<*, *>?,
            seed: () -> Unit = ::toFirestore,
        ) {
            if (!exists) {
                if (!fromCache) seed()
                return
            }
            fromFirestore(lists ?: emptyMap<String, List<Long>>(), info)
        }

        init {
            migrateOldFavorites()
        }
    }
}

File: iOS/tagmaster/tagmaster/ListModel.swift
//
//  ListModel.swift
//  tagmaster
//
//  Created by David Poll on 2025.
//  Copyright © 2025 DepollSoft. All rights reserved.
//

import Foundation
import Firebase

/// Manages a named list of tag IDs, persisted to UserDefaults and synced to Firestore.
/// Mirrors the Android `ListModel.kt` pattern.
@objc class ListModel: NSObject {

    // MARK: - Constants

    static let listsKey = "tagmaster.lists"
    private static let oldListsKey = "depollsoft.pitchperfect.lists"

    // MARK: - Static State

    private static var modelInstances: [String: ListModel] = [:]
    private static var shouldStore = true
    private static var registration: ListenerRegistration?

    /// All list data keyed by list name, backed by UserDefaults.
    static var preferences: [String: [Int]] = {
        migrateOldListsKey()
        return UserDefaults.standard.dictionary(forKey: listsKey) as? [String: [Int]] ?? [:]
    }()

    // MARK: - Instance Properties

    let listName: String
    private(set) var ids: [Int]

    // MARK: - Init

    private init(listName: String) {
        self.listName = listName
        self.ids = ListModel.preferences[listName] ?? []
        super.init()
    }

    // MARK: - Factory

    /// Returns a cached `ListModel` for the given list name, creating one if needed.
    @objc static func get(_ listName: String) -> ListModel {
        if let existing = modelInstances[listName] {
            return existing
        }
        let model = ListModel(listName: listName)
        modelInstances[listName] = model
        return model
    }

    // MARK: - Mutators

    @objc func add(_ id: Int) {
        guard !ids.contains(id) else { return }
        ids.append(id)
        storeValue()
    }

    @objc func remove(_ id: Int) {
        ids.removeAll { $0 == id }
        storeValue()
    }

    @objc func contains(_ id: Int) -> Bool {
        return ids.contains(id)
    }

    @objc func moveUp(_ id: Int) {
        guard let index = ids.firstIndex(of: id), index > 0 else { return }
        ids.remove(at: index)
        ids.insert(id, at: index - 1)
        storeValue()
    }

    @objc func moveDown(_ id: Int) {
        guard let index = ids.firstIndex(of: id), index < ids.count - 1 else { return }
        ids.remove(at: index)
        ids.insert(id, at: index + 1)
        storeValue()
    }

    @objc func canMoveUp(_ id: Int) -> Bool {
        guard let index = ids.firstIndex(of: id) else { return false }
        return index > 0
    }

    @objc func canMoveDown(_ id: Int) -> Bool {
        guard let index = ids.firstIndex(of: id) else { return false }
        return index < ids.count - 1
    }

    @objc func reset() {
        ids.removeAll()
        storeValue()
    }

    /// Replaces the entire list. Used during Firestore sync.
    func setIds(_ newIds: [Int], notify: Bool = true) {
        ids = newIds
        if ids.isEmpty {
            ListModel.preferences.removeValue(forKey: listName)
        } else {
            ListModel.preferences[listName] = ids
        }
        if notify {
            NotificationCenter.default.post(name: .userDataChanged, object: nil)
        }
    }

    // MARK: - Persistence

    private func storeValue() {
        if ids.isEmpty {
            ListModel.preferences.removeValue(forKey: listName)
        } else {
            ListModel.preferences[listName] = ids
        }
        ListModel.storeToDefaults()
        NotificationCenter.default.post(name: .userDataChanged, object: nil)

        guard ListModel.shouldStore else { return }

        guard let user = Auth.auth().currentUser else { return }
        let userDoc = Firestore.firestore().document("users/\(user.uid)")
        let value: Any = ids.isEmpty ? FieldValue.delete() : ids
        userDoc.setData(
            ["lists": [listName: value]],
            mergeFields: ["lists.\(listName)"]
        )
    }

    private static func storeToDefaults() {
        UserDefaults.standard.set(preferences, forKey: listsKey)
    }

    // MARK: - Migration

    /// Migrates from the old PitchPerfect lists key to the new TagMaster key.
    private static func migrateOldListsKey() {
        if UserDefaults.standard.dictionary(forKey: listsKey) != nil {
            // New key already exists; clean up old key if present
            UserDefaults.standard.removeObject(forKey: oldListsKey)
            return
        }
        if let oldData = UserDefaults.standard.dictionary(forKey: oldListsKey) as? [String: [Int]] {
            UserDefaults.standard.set(oldData, forKey: listsKey)
            UserDefaults.standard.removeObject(forKey: oldListsKey)
        }
    }

    // MARK: - Delete

    /// Deletes a list from local preferences. Firestore deletion is handled by CustomListsModel.
    static func deleteList(_ key: String) {
        let model = modelInstances[key]
        model?.ids.removeAll()
        preferences.removeValue(forKey: key)
        modelInstances.removeValue(forKey: key)
        storeToDefaults()
    }

    // MARK: - Test Mode

    @objc static func setTestMode(_ enabled: Bool) {
        shouldStore = !enabled
    }

    /// Clears cached model instances and preferences. For testing only.
    static func resetModelInstances() {
        modelInstances.removeAll()
        preferences.removeAll()
    }

    // MARK: - Firestore Sync

    /// Sets up a Firestore snapshot listener for the current user's document.
    @objc static func connectToFirestore() {
        registration?.remove()
        guard let user = Auth.auth().currentUser else { return }
        let userDoc = Firestore.firestore().document("users/\(user.uid)")
        registration = userDoc.addSnapshotListener { snapshot, error in
            if let error = error {
                print(error)
                return
            }
            guard let snapshot = snapshot else { return }

            if !snapshot.exists {
                // No existing user document — initialize it
                toFirestore()
                return
            }

            fromFirestore(snapshot.get("lists") as? [String: Any] ?? [:])
            CustomListsModel.fromFirestore(snapshot.get("listMeta") as? [String: Any])
        }
    }

    /// Applies Firestore list data to local state without writing back.
    static func fromFirestore(_ data: [String: Any]) {
        shouldStore = false
        defer {
            shouldStore = true
            storeToDefaults()
        }

        // Remove lists that are no longer in Firestore data
        let remoteKeys = Set(data.keys)
        for key in preferences.keys where !remoteKeys.contains(key) {
            let model = ListModel.get(key)
            model.ids.removeAll()
            preferences.removeValue(forKey: key)
        }

        // Update/add lists from Firestore data
        for (key, value) in data {
            let model = ListModel.get(key)
            let newIds: [Int]
            if let intArray = value as? [Int] {
                newIds = intArray
            } else if let numArray = value as? [NSNumber] {
                newIds = numArray.map { $0.intValue }
            } else {
                newIds = []
            }
            if newIds != model.ids {
                model.ids = newIds
                if newIds.isEmpty {
                    preferences.removeValue(forKey: key)
                } else {
                    preferences[key] = newIds
                }
            }
        }

        NotificationCenter.default.post(name: .userDataChanged, object: nil)
    }

    /// Writes all list data (and list metadata) to Firestore.
    static func toFirestore() {
        guard let user = Auth.auth().currentUser else { return }
        let userDoc = Firestore.firestore().document("users/\(user.uid)")
        var data: [String: Any] = ["lists": preferences]
        let listMetaMap = CustomListsModel.toFirestoreMap()
        if !listMetaMap.isEmpty {
            data["listMeta"] = listMetaMap
        }
        userDoc.setData(data, merge: true)
    }

    /// Removes the Firestore snapshot listener.
    @objc static func disconnectFromFirestore() {
        registration?.remove()
        registration = nil
    }
}
